src/NuGet.Services.AzureSearch/IndexActionBuilder.cs

```diff
@@ -0,0 +1,215 @@
+            metadata.SortableTitle = package.Title ?? packageId;
+            metadata.IsLatestStableSemVer1 = changes.LatestStableSemVer1;
+            metadata.IsLatestSemVer1 = changes.LatestSemVer1;
+            metadata.IsLatestStableSemVer2 = changes.LatestStableSemVer2;
+            metadata.IsLatestSemVer2 = changes.LatestSemVer2;
+
+            return IndexAction.Upload<KeyedDocument>(metadata);
+        }
+
+        private static void ApplyMetadata(IBaseMetadataDocument metadata, string packageId, Package package)
+        {
+            var parsedVersion = NuGetVersion.Parse(package.Version);
+            Guard.Assert(
+                package.NormalizedVersion == null
+                || package.NormalizedVersion == parsedVersion.ToNormalizedString(),
+                $"The calculated {nameof(Package.NormalizedVersion)} does not match the DB value for {packageId} {package.Version}.");
+            Guard.Assert(
+                package.IsPrerelease == parsedVersion.IsPrerelease,
+                $"The calculated {nameof(Package.IsPrerelease)} does not match the DB value for {packageId} {package.Version}.");
+
+            metadata.Authors = package.FlattenedAuthors;
+            metadata.Copyright = package.Copyright;
+            metadata.Created = AssumeUtc(package.Created);
+            metadata.Description = package.Description;
+            metadata.FileSize = package.PackageFileSize;
+            metadata.FlattenedDependencies = package.FlattenedDependencies;
+            metadata.Hash = package.Hash;
+            metadata.HashAlgorithm = package.HashAlgorithm;
+            metadata.IconUrl = package.IconUrl;
+            metadata.Language = package.Language;
+            metadata.LastEdited = AssumeUtc(package.Created);
+            metadata.LicenseUrl = package.LicenseUrl;
+            metadata.MinClientVersion = package.MinClientVersion;
+            metadata.NormalizedVersion = package.NormalizedVersion;
+            metadata.OriginalVersion = package.Version;
+            metadata.PackageId = packageId;
+            metadata.Prerelease = package.IsPrerelease;
+            metadata.ProjectUrl = package.ProjectUrl;
+            metadata.Published = package.Listed ? AssumeUtc(package.Published) : UnlistedPublished;
+            metadata.ReleaseNotes = package.ReleaseNotes;
+            metadata.RequiresLicenseAcceptance = package.RequiresLicenseAcceptance;
+            metadata.SemVerLevel = package.SemVerLevelKey;
+            metadata.Summary = package.Summary;
+            metadata.Tags = Utils.SplitTags(package.Tags ?? string.Empty);
+            metadata.Title = package.Title;
+        }
+
+        private static DateTimeOffset AssumeUtc(DateTime dateTime)
+        {
+            return new DateTimeOffset(
+                DateTime.SpecifyKind(dateTime, DateTimeKind.Utc),
+                TimeSpan.Zero);
+        }
+
+        private static string GetSearchDocumentKey(string packageId, SearchFilters searchFilters)
```

Comment:
I made these keys a little longer than what is minimally required so that they are a bit readable. Take a look at the test cases below for examples.

In general, I combine a "readable" string, which is not necessarily unique due to character restrictions, and a "unique" string, which is unique via base64 encoding of a unique string.